SimPEG/potential_fields/magnetics/simulation.py

```diff
@@ -247,22 +247,32 @@ def evaluate_integral(self, receiver_location, components):
             Tz = [Tzx Tzy Tzz]
         """
         # TODO: This should probably be converted to C
-        eps = 1e-8  # add a small value to the locations to avoid /0
+        tol1 = 1e-10    # Tolerance 1 for numerical stability over nodes and edges
+        tol2 = 1e-4     # Tolerance 2 for numerical stability over nodes and edges
 
         rows = {component: np.zeros(3 * self.Xn.shape[0]) for component in components}
 
         # number of cells in mesh
         nC = self.Xn.shape[0]
 
+        # base cell dimensions
+        min_hx, min_hy, min_hz = self.mesh.hx.min(), self.mesh.hy.min(), self.mesh.hz.min()
+
         # comp. pos. differences for tne, bsw nodes
-        dz2 = self.Zn[:, 1] - receiver_location[2] + eps
-        dz1 = self.Zn[:, 0] - receiver_location[2] + eps
+        dz2 = self.Zn[:, 1] - receiver_location[2]
+        dz2[np.abs(dz2)/min_hz < tol2] = tol2 * min_hz
```

Comment:
I am not a huge fan of this, because indexing with logical will likely hurt us when parallelizing. Better if we can stick we adding small constants.

Reply:
It really isn't stable though. The errors above nodes and especially edges are too large. But we can try to figure out a way to accomplish both goals.

Reply:
Just to leave a trace, @dccowan ran it both ways on his machine on a 500 x 200k size problem. No significant change in runtime, but definitely an improvement in accuracy. Well done.